jupyter-extension/src/plugins/mount/mountDrive.ts

```diff
@@ -1,61 +1,123 @@
+import {Paging} from './paging';
+
+const MAX_NUM_CONTENTS_PAGE = 100;
+const DEFAULT_CONTENT_MODEL: Contents.IModel = {
+  name: '',
+  path: '',
+  last_modified: '',
+  created: '',
+  content: [],
+  format: 'json',
+  mimetype: '',
+  size: undefined,
+  writable: true,
+  type: 'directory',
+};
 
 export class MountDrive implements Contents.IDrive {
   public _registry: DocumentRegistry;
+  readonly _model: Paging.IModel;
   private _fileChanged = new Signal<this, Contents.IChangedArgs>(this);
+  private _loading = new Signal<this, boolean>(this);
   private _serverSettings = ServerConnection.makeSettings();
   private _isDisposed = false;
+  private _cache: {key: string | null; contents: any};
   modelDBFactory?: ModelDB.IFactory | undefined;
 
   constructor(registry: DocumentRegistry) {
     this._registry = registry;
+    this._model = {page: 0, max_page: 0};
+    this._cache = {key: null, contents: DEFAULT_CONTENT_MODEL};
   }
 
   get name(): 'mount-browser' {
     return 'mount-browser';
   }
+
+  get model(): Paging.IModel {
+    return this._model;
+  }
+
   get fileChanged(): ISignal<this, Contents.IChangedArgs> {
     return this._fileChanged;
   }
 
+  // Signal emits `true` on expensive request to backend and `false` on response.
+  get loading(): ISignal<this, boolean> {
+    return this._loading;
+  }
+
   get serverSettings(): ServerConnection.ISettings {
     return this._serverSettings;
   }
   get isDisposed(): boolean {
     return this._isDisposed;
   }
 
+  async _get(
+    url: string,
+    options: PartialJSONObject,
+  ): Promise<Contents.IModel> {
+    url += URLExt.objectToQueryString(options);
+    return requestAPI<Contents.IModel>(url, 'GET');
+  }
+
+  // We might be able to pass pagination to the ContentsManager backend by including
+  // it in the URI specifier. This would require a custom ContentsManager which
+  // supports pagination.
   async get(
     localPath: string,
     options?: Contents.IFetchOptions,
   ): Promise<Contents.IModel> {
+    const url = URLExt.join('pfs', localPath);
+
+    // Make a no-content request to determine the type of content being requested.
+    // If the content type is not a directory, then we want to preserve our cache
+    //   and not paginate the results.
+    let shallowResponse;
     try {
-      const response = await requestAPI<Contents.IModel>(
-        URLExt.join('pfs', localPath),
-        'GET',
-      );
-      return response;
+      shallowResponse = await this._get(url, {content: '0'});
     } catch (e) {
       console.log('/pfs not found');
-      return {
-        name: '',
-        path: '',
-        last_modified: '2022-04-26T16:28:48.015858Z',
-        created: '2022-04-26T16:28:48.015858Z',
-        content: [],
-        format: 'json',
-        mimetype: '',
-        size: undefined,
-        writable: true,
-        type: 'directory',
-      };
+      return DEFAULT_CONTENT_MODEL;
+    }
+    const content = options?.content ? '1' : '0';
```

Comment:
Is this checking for the existence of the field `content` or checking if its value is truthy? Can we make the check more explicit?

Reply:
The ContentManager API expects the `content` value to either be `'1'` or `'0'` where `1` stands for "send back content" and `0` is the opposite. This is translating the `true|false` value into `'1'|'0'` and defaulting to `'0'`. I've pulled this line straight from the `jupterlab` source: [link](https://github.com/jupyterlab/jupyterlab/blob/1728207ffbad5baea1e77e4005230632bf6f16a6/packages/services/src/contents/index.ts#L1124)

Surprisingly, we weren't setting the `content` value at all before this change and it still worked, so the ContentsManager must default to always sending back content. From my testing it appears that at start up file browser makes several no-content calls to backend, so this likely contributed to some of the slowness.